### ISSUE TITLE
Bundling with webpack will fail by default when not targeting browser

### ISSUE THREAD

**Issue by Domiii** [websockets/ws#1683]
Due to #1538, webpack builds will fail by default ([workaround provided](https://github.com/websockets/ws/issues/1538#issuecomment-504048708)).

Considering how popular webpack is, would be great if there can be a more versatile solution. As [this guy](https://github.com/webpack/webpack/issues/7646) points out, it is easily fixed by just not using the `package.json` `browser` field anymore. Any thoughts?

(PS: I just spent hours debugging [this `engine.io` bug](https://github.com/socketio/engine.io/issues/595) not allowing me to connect, because they accidentally silenced the error. - Not saying that this is your fault, just referencing.)

**Comment by lpinca** [comment]
I'm 👎 on removing the `browser` field because if we do that people will start opening issues again about not being able to bundle for the browser like it was done countless of times in the past despite it being documented https://github.com/websockets/ws#ws-a-nodejs-websocket-library.

**Comment by Domiii** [comment]
@lpinca You can just add an `if` in `index.js` instead - or have `browser.js` straight-up raise a syntax error, combined with a meanigful error message, rather than letting them require/bundle it and waiting until way later to let them know?

If the consensus is to not change the code, I would vote in favor of at least amending the `README` with a `Common Problems` section, specifically getting into the broken webpack default behavior of targeting `browser` even for a `node` or `umd` build? (Probably need to knock on `socket.io`'s door as well, as that is probably one of your bigger consumers)

**Comment by lpinca** [comment]
To be honest I think everything work as expected here:

- Bundlers default target is the browser. It has always been like this. I think it makes sense, and I don't expect it to be changed any time soon.
- When the bundle is built for the browser an error is thrown. We could throw the error when the module is loaded instead of throwing it when calling the constructor but I think it does not make a big difference as the bundle would be generated anyway. The error will only be thrown at run time.
- The bundler should be configured to produce a bundle for a specific environment. webpack has the [`target`](https://webpack.js.org/configuration/target/) option for this.

**Comment by Domiii** [comment]
@lpinca Ok, I can agree with that. Then what about sharing a working webpack sample and refering to it in the docs/in the README?

I specifically would recommend it, because:

1. (obviously) I myself battled with this for quite a while (so I'm sure that the next person might stumble into the same pitfalls), and...
1. I wasn't  even aware of the `package.json` `browser` entry point
1. Despite all the time I invested, I only just now heard of the fact that `target` affects the chosen `package.json` entry point. It seems that this relationship is still undocumented: I cannot see that explained [in the relevant webpack docs](https://webpack.js.org/concepts/targets/), and [here](https://stackoverflow.com/a/39176835) it is claimed to be undocumented.
1. It is entirely nonintuitive that it chooses `browser` if my `libraryTarget` is already set to `umd`

At the end of the day, I guess this is a bit of me ranting on webpack's nonintuitive default behavior and confusing configuration, but still... I think my point is valid :)

**Comment by Domiii** [comment]
@lpinca Ok I agree. Might want to mention it in the `README` though :)

(PS: I deleted my previous comment which was too much of a rant. Also thanks for pointing out the `target` option. Turns out that it itself is not properly documented, however light is shone on it in the [resolve.mainFields](https://webpack.js.org/configuration/resolve/#resolvemainfields) documentation, which ironically I looked at before; must have been blind not to see the explanation there.)
